lib/build-spec.ts

```diff
@@ -0,0 +1,212 @@
+        paths: mergeList(a.paths, b.paths)!
+      }))
+    });
+
+    function mergeArtifacts(a: PrimaryArtifactStruct, b: PrimaryArtifactStruct): PrimaryArtifactStruct {
+      if (a.files || b.files) {
+        throw new Error('None of the BuildSpecs may have a primary artifact.');
+      }
+
+      const artifacts = Object.assign({}, a["secondary-artifacts"] || {});
+      for (const [k, v] of Object.entries(b["secondary-artifacts"] || {})) {
+        if (k in artifacts) {
+          throw new Error(`There is already an artifact with name ${k}`);
+        }
+        artifacts[k] = v;
+      }
+      return Object.assign({}, a, { 'secondary-artifacts': artifacts });
+    }
+
+    function equalObjects(a: string, b: string) {
+      if (a !== b) {
+        throw new Error(`Can't merge two different values for the same key: ${JSON.stringify(a)}, ${JSON.stringify(b)}`);
+      }
+      return b;
+    }
+
+    function mergeObj<T>(a: T | undefined, b: T | undefined, fn: (a: T, b: T) => T): T | undefined {
+      if (a === undefined) { return b; }
+      if (b === undefined) { return a; }
+      return fn(a, b);
+    }
+
+    function mergeDict<T>(as: {[k: string]: T} | undefined, bs: {[k: string]: T} | undefined, fn: (a: T, b: T) => T) {
+      return mergeObj(as, bs, (a, b) => {
+        const ret = Object.assign({}, a);
+        for (const [k, v] of Object.entries(b)) {
+          if (ret[k]) {
+            ret[k] = fn(ret[k], v);
+          } else {
+            ret[k] = v;
+          }
+        }
+        return ret;
+      });
+    }
+
+    function mergeList<T>(as: T[] | undefined, bs: T[] | undefined): T[] | undefined {
+      return mergeObj(as, bs, (a, b) => a.concat(b));
+    }
+  }
+
+  public render(options: BuildSpecRenderOptions = {}): BuildSpecStruct {
+    return Object.assign({}, this.spec, { artifacts: this.renderArtifacts(options) });
+  }
+
+  private renderArtifacts(options: BuildSpecRenderOptions): PrimaryArtifactStruct | undefined {
+    if (!this.spec.artifacts || !this.spec.artifacts['secondary-artifacts']) { return this.spec.artifacts; }
+
+    // Simplify a single "secondary-artifacts" to a single primary artifact (regardless of the name)
+    const singleArt = dictSingletonValue(this.spec.artifacts['secondary-artifacts']);
+    if (singleArt) { return singleArt; }
+
+    // Otherwise rename a 'PRIMARY' key if it exists
+    if (MAGIC_ARTIFACT_NAME in this.spec.artifacts['secondary-artifacts']) {
+      if (!options.primaryArtifactName) {
+        throw new Error(`Replacement name for ${MAGIC_ARTIFACT_NAME} artifact not supplied`);
+      }
+
+      return { 'secondary-artifacts': renameKey(this.spec.artifacts['secondary-artifacts'], MAGIC_ARTIFACT_NAME, options.primaryArtifactName) };
+    }
+
+    return this.spec.artifacts;
+  }
+}
+
+export interface SimpleBuildSpecProps {
+  preBuild?: string[];
+  build?: string[];
+
+  /**
+   * Where the directories for each artifact are
+   *
```

Comment:
I kind of liked the previous API better that optimized for a single primary artifact:

```ts
interface SimpleBuildSpecProps {
  // ...
  artifactsDir: string;

  additionalArtifactDirs? : { [ ... ] }
}